spring-cloud-stream-binder-kafka/src/main/java/org/springframework/cloud/stream/binder/kafka/KafkaBinderMetrics.java

```diff
@@ -53,72 +61,93 @@
+			List<TopicPartition> topicPartitions = new LinkedList<>();
+			for (PartitionInfo partitionInfo : partitionInfos) {
+				topicPartitions.add(new TopicPartition(partitionInfo.topic(), partitionInfo.partition()));
+			}
+			Map<TopicPartition, Long> endOffsets = metadataConsumer.endOffsets(topicPartitions);
+
+			for (Map.Entry<TopicPartition, Long> endOffset : endOffsets.entrySet()) {
+				OffsetAndMetadata current = metadataConsumer.committed(endOffset.getKey());
+				if (current != null) {
+					lag += endOffset.getValue() - current.offset();
 				}
-				Map<TopicPartition, Long> endOffsets = metadataConsumer.endOffsets(topicPartitions);
-				long lag = 0;
-				for (Map.Entry<TopicPartition, Long> endOffset : endOffsets.entrySet()) {
-					OffsetAndMetadata current = metadataConsumer.committed(endOffset.getKey());
-					if (current != null) {
-						lag += endOffset.getValue() - current.offset();
-					}
-					else {
-						lag += endOffset.getValue();
-					}
+				else {
+					lag += endOffset.getValue();
 				}
-				registry.gauge(String.format("%s.%s.%s.lag", METRIC_PREFIX, group, topic), lag);
-			}
-			catch (Exception e) {
-				LOG.debug("Cannot generate metric for topic: " + topic, e);
 			}
 		}
+		catch (Exception e) {
+			LOG.debug("Cannot generate metric for topic: " + topic, e);
+		}
+		return lag;
 	}
 
 	private ConsumerFactory<?, ?> createConsumerFactory(String group) {
-		if (defaultConsumerFactory != null) {
-			return defaultConsumerFactory;
-		}
-		Map<String, Object> props = new HashMap<>();
-		props.put(ConsumerConfig.KEY_DESERIALIZER_CLASS_CONFIG, ByteArrayDeserializer.class);
-		props.put(ConsumerConfig.VALUE_DESERIALIZER_CLASS_CONFIG, ByteArrayDeserializer.class);
-		if (!ObjectUtils.isEmpty(binderConfigurationProperties.getConsumerConfiguration())) {
-			props.putAll(binderConfigurationProperties.getConsumerConfiguration());
+		if (this.defaultConsumerFactory == null) {
+			Map<String, Object> props = new HashMap<>();
+			props.put(ConsumerConfig.KEY_DESERIALIZER_CLASS_CONFIG, ByteArrayDeserializer.class);
+			props.put(ConsumerConfig.VALUE_DESERIALIZER_CLASS_CONFIG, ByteArrayDeserializer.class);
+			if (!ObjectUtils.isEmpty(binderConfigurationProperties.getConsumerConfiguration())) {
+				props.putAll(binderConfigurationProperties.getConsumerConfiguration());
+			}
+			if (!props.containsKey(ConsumerConfig.BOOTSTRAP_SERVERS_CONFIG)) {
+				props.put(ConsumerConfig.BOOTSTRAP_SERVERS_CONFIG,
+						this.binderConfigurationProperties.getKafkaConnectionString());
+			}
+			props.put("group.id", group);
+			this.defaultConsumerFactory = new DefaultKafkaConsumerFactory<>(props);
 		}
-		if (!props.containsKey(ConsumerConfig.BOOTSTRAP_SERVERS_CONFIG)) {
-			props.put(ConsumerConfig.BOOTSTRAP_SERVERS_CONFIG,
-					this.binderConfigurationProperties.getKafkaConnectionString());
+
+		return this.defaultConsumerFactory;
+	}
+	
+	@Override
+	public void onApplicationEvent(BindingCreatedEvent event) {
```

Comment:
Right, the event is here because you publish it from the Binder per se, into its own child context, not to the parent.